edgraph/access_ee_test.go

```diff
@@ -0,0 +1,112 @@
+package edgraph
+
+import (
+	"testing"
+	"time"
+
+	"github.com/dgraph-io/dgraph/ee/acl"
+	"github.com/dgraph-io/dgraph/worker"
+
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+func generateJWT(namespace uint64, userId string, groupIds []string, expiry int64) string {
+	claims := jwt.MapClaims{"namespace": namespace, "userid": userId, "exp": expiry}
+	if groupIds != nil {
+			claims["groups"] = groupIds
+	}
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, &claims)
+
+	tokenString, _ := token.SignedString([]byte(worker.Config.HmacSecret))
+
+	return tokenString
+}
+
+func sliceCompare(x, y []string) bool {
+	if len(x) != len(y) {
+		return false
+	}
+
+	for i := range x {
+		if x[i] != y[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func TestValidateToken(t *testing.T) {
+	expiry := time.Now().Add(time.Minute * 30).Unix()
+	userDataList := []userData{
+		{1234567890, "user1", []string{"701", "702"}},
+		{2345678901, "user2", []string{"703", "701"}},
+		{3456789012, "user3", []string{"702", "703"}},
+	}
+
+	for _, userdata := range userDataList {
+		tokenString := generateJWT(userdata.namespace, userdata.userId, userdata.groupIds, expiry)
+
+		ud, _ := validateToken(tokenString)
```

Comment:
I suspect that you have a local change, as otherwise, at line 51 below, we should expect an error since ud would be nil.
See: https://github.com/dgraph-io/dgraph/blob/bc71dc3297880e8c282be638506eddab9270b010/edgraph/access.go#L93

Reply:
edgraph/access.go and edgraph/access_ee.go declares "func validateToken(jwtStr string) ([]string, error)" and "func validateToken(jwtStr string) (*userData, error)" respectively.

The validateToken(tokenString) call from edgraph/access_ee_test.go resolves to "func validateToken(jwtStr string) (*userData, error)" declared in edgraph/access_ee.go but not to "func validateToken(jwtStr string) ([]string, error)" declared in edgraph/access.go, at least, in my linux workspace. The test output shown below. The text Italic bold is the result of an extra statement I have added to the test code, 'fmt.Println (ud)', for verification.

jassi Wed Dec 14 11:58:08 edgraph_validatetoken_getaccessjwt_getrefreshjwt edgraph > go test -v -count=1 ./...
=== RUN   TestValidateToken
**_&{1234567890 user1 [701 702]}
&{2345678901 user2 [703 701]}
&{3456789012 user3 [702 703]}_**
--- PASS: TestValidateToken (0.00s)
=== RUN   TestGetAccessJwt
--- PASS: TestGetAccessJwt (0.00s)
=== RUN   TestGetRefreshJwt
--- PASS: TestGetRefreshJwt (0.00s)
=== RUN   TestParseNQuads
--- PASS: TestParseNQuads (0.00s)
=== RUN   TestValNquads
. . .

When I write a sample program with an Add() function I see a different behaviour altogether, go interpreter throws a message- 'Add redeclared', whereas this is not seen for edgraph/ above.
This is the sample program code-

jassi Wed Dec 14 12:11:06 edgraph_validatetoken_getaccessjwt_getrefreshjwt math > ls -l
total 12
-rw-r--r-- 1 jassi users  65 Dec 14 07:56 m.go
-rw-r--r-- 1 jassi users 148 Dec 14 08:04 m_1.go
-rw-r--r-- 1 jassi users 515 Dec 14 10:45 m_1_test.go
jassi Wed Dec 14 12:11:08 edgraph_validatetoken_getaccessjwt_getrefreshjwt math > 
jassi Wed Dec 14 12:11:09 edgraph_validatetoken_getaccessjwt_getrefreshjwt math > cat m.go
package math

func Add (x int, y int) ([]string) {
	return nil
}
jassi Wed Dec 14 12:11:15 edgraph_validatetoken_getaccessjwt_getrefreshjwt math > cat m_1.go
package math

type userData struct {
	name string
	uid int64
}

func Add (x int, y int) (*userData) {
	return &userData{name: "Hello", uid: 2233}
}
jassi Wed Dec 14 12:11:25 edgraph_validatetoken_getaccessjwt_getrefreshjwt math >
jassi Wed Dec 14 12:11:33 edgraph_validatetoken_getaccessjwt_getrefreshjwt math > cat m_1_test.go
package math

import (
	"testing"
)

func Test_Add (t *testing.T) {
	Add (3, 5)
/*
	got := Add (3, 5)
	want := 8

	if got != want {
		t.Errorf ("got %v, wanted %v", got, want)
	}
*/
}
jassi Wed Dec 14 12:12:03 edgraph_validatetoken_getaccessjwt_getrefreshjwt math > go test -v -count=1 ./...
# github.com/dgraph-io/dgraph/gotest/math [github.com/dgraph-io/dgraph/gotest/math.test]
./m_1.go:8:6: Add redeclared in this block
	./m.go:3:6: other declaration of Add
FAIL	github.com/dgraph-io/dgraph/gotest/math [build failed]
FAIL
jassi Wed Dec 14 12:12:25 edgraph_validatetoken_getaccessjwt_getrefreshjwt math >

I don't know what is causing this discrepancy.